### PR TITLE
sink(cdc): fix the bug that mq sink can lost callbacks

### LINKED ISSUE #9855: TiCDC MQ sinks can lost callbacks
### What did you do?

TiCDC MQ sinks can lost callbacks, which makes checkpoint can be stuck forever.

### What did you expect to see?

_No response_

### What did you see instead?

TiCDC MQ sinks can lost callbacks, which makes checkpoint can be stuck forever.

And, it's the root cause of #9542 .

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
master
```

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Issue Number: close #9855 

### What is changed and how it works?

When errors happen in `MQSink.WriteEvents`, the backend sink should be canceled correctly. Otherwise the table can lost callbacks forever.

#### Tests 

 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by 3AceShowHand — APPROVED** [review]

File: cdc/sink/dmlsink/mq/mq_dml_sink.go [hunk 1]
```diff
@@ -144,10 +147,16 @@ func (s *dmlSink) WriteEvents(txns ...*dmlsink.CallbackableEvent[*model.SingleTa
...
+
 		for _, row := range txn.Event.Rows {
 			topic := s.alive.eventRouter.GetTopicForRowChange(row)
 			partitionNum, err := s.alive.topicManager.GetPartitionNum(s.ctx, topic)
+			failpoint.Inject("MQSinkGetPartitionError", func() {
+				log.Info("failpoint MQSinkGetPartitionError injected", zap.String("changefeedID", s.id.ID))
+				err = errors.New("MQSinkGetPartitionError")
+			})
 			if err != nil {
+				s.cancel(err)
```
Comment (CharlesCheung96) [inline cdc/sink/dmlsink/mq/mq_dml_sink.go:159]: Why could this cancel prevent callback from being lost?

Comment (CharlesCheung96) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:159]: In addition, when WriteEvents returns error, the external caller should also close dml_sink?

Comment (hicqu) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:159]: Here we cancel the sink backend. Then when checking whether the table sink has been closed or not, we can know the sink backend is dead, and then we don't need to be blocked on table sink callbacks.

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**